csharp/src/Ice/Connection.cs

```diff
@@ -1180,27 +1191,30 @@ private async Task PerformCloseAsync()
                 _communicator.Logger.Error("unexpected connection exception:\n" + ex + "\n" + _transceiver.ToString());
             }
 
-            // Notify pending requests of the failure
-            foreach (TaskCompletionSource<IncomingResponseFrame> response in _requests.Values)
+            // Notify pending requests of the failure and the close callback. We use the thread pool to ensure the
+            // continuations or the callback are not run from this thread which might still lock the connection's mutex.
+            _ = Task.Run(() =>
             {
-                response.SetException(_exception!);
-            }
-            _requests.Clear();
+                foreach (TaskCompletionSource<IncomingResponseFrame> response in _requests.Values)
```

Comment:
Could we pick a better variable name than `response` for the TaskCompletionSource?

Reply:
Renamed it `responseTaskCompletionSource`.